keps/sig-auth/3325-self-user-attributes-review-api/README.md

```diff
@@ -0,0 +1,357 @@
+The user will hip the endpoint after authentication happens, so all attributes will be available to return.
+
+## Design Details
+
+This design is inspired by the `*AccessReview` and `TokenReview` APIs.
+The endpoint has no input parameters or a `spec` field because only the authentication result is required.
+
+### Request
+
+The structure for building a request:
+```go
+type SelfUserAttributesReview struct {
+	metav1.TypeMeta `json:",inline"`
+	// Standard list metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#metadata
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+	// Status is filled in by the server with the user attributes.
+	Status SelfUserAttributesReviewStatus `json:"status,omitempty" protobuf:"bytes,2,opt,name=status"`
+}
+```
+```go
+type SelfUserAttributesReviewStatus struct {
+	// User attributes of the current user.
+	// +optional
+	UserInfo UserInfo `json:"userInfo,omitempty" protobuf:"bytes,1,opt,name=userInfo"`
+}
+
+type UserInfo struct {
+	Name   string              `json:"name" protobuf:"bytes,1,opt,name=name"`
+	UID    string              `json:"uid" protobuf:"bytes,2,opt,name=uid"`
+	Groups []string            `json:"groups" protobuf:"bytes,1,opt,name=groups"`
+	Extra  map[string][]string `json:"extra" protobuf:"bytes,1,opt,name=extra"`
+}
+```
+
+On receiving a request, the Kubernetes API server fills the status with the user attributes and returns it to the user.
+
+Request URL:
+```
+GET /apis/authentication.k8s.io/v1alpha1/selfuserattributesreview
+```
+Response example:
+
+```json
+{
+  "apiVersion": "authentication.k8s.io/v1alpha1",
+  "kind": "SelfUserAttributesReview",
+  "status": {
+    "name": "jane.doe",
+    "uid": "b6c7cfd4-f166-11ec-8ea0-0242ac120002",
+    "groups": ["viewers", "editors"],
+    "extra": {
+      "provider_id": "token.company.dev"
+    }
+  }
+}
+```
+
+User attributes are known at the moment of accessing the rest API endpoint and can be extracted from the request context.
+
+### RBAC
+
+RBAC rules to grant access to this API should be present in the cluster by default.
+It is implied that the `system:basic-user` cluster role will be extended to the following:
+```yaml
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRole
+metadata:
+  annotations:
+    rbac.authorization.kubernetes.io/autoupdate: "true"
+  creationTimestamp: null
+  labels:
+    kubernetes.io/bootstrapping: rbac-defaults
+  name: system:basic-user
+rules:
+- apiGroups:
+  - authorization.k8s.io
+  resources:
+  - selfsubjectaccessreviews
+  - selfsubjectrulesreviews
+  verbs:
+  - create
+- apiGroups:
+  - authentication.k8s.io
+  resources:
+  - selfuserattributesreview
+  verbs:
+  - create
+```
+
+This API is enabled by default and can be disabled by the following kube-apiserver flag (along with the TokenReview API).
+```
+--runtime-config=authentication.k8s.io/v1=false
+```
```

Comment:
Disabling token review isn't practical?  I am fine with this not being part of conformance but I denying this via a validating admission webhook seems more practical (which is possible since this is a `create`).

@deads2k you mentioned:

> the command line flag used to disable this API if a cluster-admin wants it off

Is that requesting a specific flag that disables this one API?

Reply:
The note about using validating webhook was added. Thanks for the hint.

Reply:
There is the KEP https://github.com/kubernetes/enhancements/pull/3137 that extended the runtime-config option to be able to disable a single resource of an API like this:
```
--runtime-config=authentication.k8s.io/v1alpha1/selfsubjectattributesreview=false
```

Added a note about it to the KEP.